<a href="https://colab.research.google.com/github/vavilon5/Educational/blob/main/numerical_methods/lab_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Лабораторная работа №1
##по теме «Методы решения нелинейных уравнений»
##Васильев В. Вариант 8

###Задание: методом итераций и методом Ньютона решить следующее НЛУ: $$cos(x)–\sqrt{x + 2} + 1 = 0$$

##Этап отделения корней

In [ ]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go

from sympy import *

In [ ]:
def f(x):
    return np.cos(x) - np.sqrt(x+2) + 1

In [ ]:
begin = -2
end = 7
step = 0.2
size = 1000
x = np.linspace(begin, end, size)
y_f = f(x)

In [ ]:
fig = go.Figure() # layout_xaxis_range = [-2,6]
fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))

fig.add_trace(go.Scatter(x=x,
                         y=y_f,
                         name="График функции"))

fig.update_layout(title="Построение графика функции",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции",
                  margin=dict(l=10, r=70, t=50, b=0))

fig.show()

Из построенного графика функции $f(x)$ видно, что на отрезке $(0, 1.5)$ есть один корень.

In [ ]:
# Вычислим производную функции f(z)
z = Symbol('z')

f_ = cos(z) - sqrt(z+2) + 1
f_1derivative = f_.diff(z)

f1 = lambdify(z, f_1derivative)

f2 = lambdify(z, f_1derivative.diff(z))


In [ ]:
f_1derivative

-sin(z) - 1/(2*sqrt(z + 2))

In [ ]:
a = 0
b = 1.5
begin = a - 1
end = b + 1
size = 1000
x = np.linspace(begin, end, size)

In [ ]:
x_temp = np.arange(a, b+0.1, 0.1)
df = pd.DataFrame(data={'x': x_temp,
                        'f(x)': f(x_temp),
                        'f1(x)': f1(x_temp),
                        'f2(x)': f2(x_temp)})
df

x      f(x)     f1(x)     f2(x)
0   0.0  0.585786 -0.353553 -0.911612
1   0.1  0.545866 -0.444866 -0.912854
2   0.2  0.496827 -0.535769 -0.903453
3   0.3  0.438761 -0.625210 -0.883665
4   0.4  0.371868 -0.712167 -0.853822
5   0.5  0.296444 -0.795653 -0.814337
6   0.6  0.212884 -0.874729 -0.765704
7   0.7  0.121675 -0.948508 -0.708492
8   0.8  0.023387 -1.016163 -0.643348
9   0.9 -0.081329 -1.076937 -0.570988
10  1.0 -0.191749 -1.130146 -0.492190
11  1.1 -0.307086 -1.175188 -0.407793
12  1.2 -0.426497 -1.211548 -0.318685
13  1.3 -0.549091 -1.238799 -0.225796
14  1.4 -0.673942 -1.256613 -0.130090
15  1.5 -0.800091 -1.264756 -0.032557

##Этап уточнения корня

###Метод итераций

Приведем уравнение $f(x)=0$ к виду $x=𝞅(x)$. Тогда рекуррентная формула $x_{n+1}=𝞅(x_{n})$. Для сходимости процесса простых итерации необходимо, чтобы $|𝞅'(x)|<1$ при $x\in [a;b]$. Если $|𝞅'(x)|>1$ то сходимость не обеспечена.
Приведем уравнение $f(x)=cos(x)–\sqrt{x + 2} + 1$ к виду $x=(cos(x)+1)^2 - 2$ и проведем исследование.


In [ ]:
# Определим функцию ф
def phi(x):
    return ((np.cos(x)+1)**2 - 2)

In [ ]:
# Вычислим первую производную функции ф
z = Symbol('z')

phi_ = ((cos(z)+1)**2 - 2)

phi_derivative = phi_.diff(z)
phi_derivative

-2*(cos(z) + 1)*sin(z)

In [ ]:
phi1 = lambdify(z, phi_derivative)

In [ ]:
y_phi = phi(x)
y_phi1 = phi1(x)

In [ ]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))

fig.add_trace(go.Scatter(x=x,
                         y=y_phi,
                         name="График функции фи"))

fig.add_trace(go.Scatter(x=[a, a],
                         y=[-2, 2],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-2, 2],
                         name="Ось B"))

fig.add_trace(go.Scatter(x=x,
                         y=x,
                         name="Ограничение сверху"))

fig.update_layout(title="Построение функции фи",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")

fig.show()

In [ ]:
# Изобразим график первой производной ф-ии ф
fig = go.Figure()
fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))

fig.add_trace(go.Scatter(x=x,
                         y=y_phi1,
                         name="График производной функции фи"))

fig.add_trace(go.Scatter(x=[a, a],
                         y=[-2, 1],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-2, 1],
                         name="Ось B"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[1, 1],
                         name="Ограничение сверху"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[-1, -1],
                         name="Ограничение снизу"))


fig.update_layout(title="Построение производной функции фи",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")

fig.show()

Как видно на графике выше $|𝞅'(x)|>1$ на отрезке $(a,b)$, поэтому построим функцию $𝞅(x) = x + \lambda f(x)$, где параметр $\lambda$ может быть определен по правилу: $\lambda = -\frac{1}{f'(\hat{x})}$, а в знаменатель требуется подставить $f'(\hat{x})$ = $\max\limits_{a \leq x \leq b}|f'(x)|$. Найдем параметр $\lambda$.

In [ ]:
# phi1(x[(x > a) & (x < b)])

In [ ]:
# Вычислим производную функции f(z)
z = Symbol('z')

f_ = cos(z) - sqrt(z+2) + 1
f_1derivative = f_.diff(z)

f1 = lambdify(z, f_1derivative)
y_f1 = f1(x)

f_1derivative

-sin(z) - 1/(2*sqrt(z + 2))

In [ ]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=x,
                         y=y_f1,
                         name="Первая производная функции"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))


fig.add_trace(go.Scatter(x=[a, a],
                         y=[-3, 2],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-3, 2],
                         name="Ось B"))

fig.update_layout(title="Построение производной функции f",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")

fig.show()

In [ ]:
# print(max(max(abs(y_f1)), min(y_f1)))
maxi = min(y_f1)
lamda = -1/maxi
lamda, maxi

(0.7903289381124969, -1.265295944228298)

Тем самым $\lambda=\frac{1}{1.265}$. Тогда рекурентная формула: $$𝞅(x)=x+\frac{cos(x)–\sqrt{x + 2} + 1}{1.265}$$

In [ ]:
def phi(x):
    return x + lamda * f(x)

In [ ]:
z = Symbol('z')

phi_ = z + lamda * f_

phi_derivative = phi_.diff(z)
phi_derivative

-0.790328938112497*sin(z) + 1 - 0.395164469056248/sqrt(z + 2)

In [ ]:
phi1 = lambdify(z, phi_derivative)

In [ ]:
y_phi = phi(x)
y_phi1 = phi1(x)

In [ ]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))

fig.add_trace(go.Scatter(x=x,
                         y=y_phi1,
                         name="График производной функции фи"))

fig.add_trace(go.Scatter(x=[a, a],
                         y=[-2, 1],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-2, 1],
                         name="Ось B"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[1, 1],
                         name="Ограничение сверху"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[-1, -1],
                         name="Ограничение снизу"))


fig.update_layout(title="Построение производной функции фи",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")

fig.show()

In [ ]:
x0 = a
sp_x = [x0]
sp_f = [f(x0)]
for i in range(3):
  x_n = phi(x0)
  sp_x.append(x_n)
  sp_f.append(f(x_n))
  x0 = x_n

In [ ]:
df = pd.DataFrame(data={'n':range(i+2), 'x': sp_x, 'f(x)': sp_f})
df

n         x      f(x)
0  0  0.000000  0.585786
1  1  0.462964  0.325349
2  2  0.720097  0.102470
3  3  0.801082  0.022287

In [ ]:
q = max(phi1(x[(x>a)&(x<b)]))
R = q/(1-q) * abs(sp_x[-1] - sp_x[-2])
R

0.20735557222388623

In [ ]:
abs(sp_x[-1] - sp_x[-2])

0.08098507504686381

## Метод Ньютона

In [ ]:
# Вычислим производные функции f(z)
z = Symbol('z')

f_ = cos(z) - sqrt(z+2) + 1
f_1derivative = f_.diff(z)

f1 = lambdify(z, f_1derivative)
y_f1 = f1(x)

f_1derivative
# dx = x[1]-x[0]
# dydx = np.gradient(y, dx)

-sin(z) - 1/(2*sqrt(z + 2))

In [ ]:
f2 = lambdify(z, f_1derivative.diff(z))
y_f2 = f2(x)
f_1derivative.diff(z)

-cos(z) + 1/(4*(z + 2)**(3/2))

In [ ]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=x,
                         y=y_f1,
                         name="Первая производная функции"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))


fig.add_trace(go.Scatter(x=[a, a],
                         y=[-3, 2],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-3, 2],
                         name="Ось B"))

fig.update_layout(title="Построение производной функции f",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")

fig.show()

In [ ]:
fig = go.Figure(layout_yaxis_range=[-4,4])

fig.add_trace(go.Scatter(x=x,
                         y=y_f2,
                         name="Вторая производная функции"))

fig.add_trace(go.Scatter(x=[begin, end],
                         y=[0, 0],
                         name="Ось Х"))


fig.add_trace(go.Scatter(x=[a, a],
                         y=[-10, 2],
                         name="Ось A"))

fig.add_trace(go.Scatter(x=[b, b],
                         y=[-10, 2],
                         name="Ось B"))

fig.update_layout(title="Построение второй производной функции f",
                  xaxis_title="Значение аргумента",
                  yaxis_title="Значение функции")


fig.show()

In [ ]:
E = [i for i in range(-2, -5, -1)]

sp_i = []
for eps in E:
  x0 = a
  i = 0
  sp_x = [x0]
  sp_f = [f(x0)]
  x_n = x0 - f(x0)/f1(x0)
  while abs(x_n - x0) > 10**eps:
    if eps == -4:
      sp_x.append(x_n)
      sp_f.append(f(x_n))
    x0 = x_n
    x_n = x0 - f(x0)/f1(x0)
    i += 1
  sp_i.append(i)


In [ ]:
df = pd.DataFrame(data={'n':range(sp_i[-1]+1), 'x': sp_x, 'f(x)': sp_f})
df

n         x          f(x)
0  0  0.000000  5.857864e-01
1  1  1.656854 -9.982421e-01
2  2  0.863191 -4.208117e-02
3  3  0.823320 -4.833379e-04
4  4  0.822851 -6.892568e-08

([-2, -3, -4], [3, 3, 4])

In [ ]:
fig = go.Figure(layout_yaxis_range = [2,5])

fig.add_trace(go.Scatter(x=E,
                         y=sp_i,
                         name="Количество итераций для погрешности"))

fig.update_layout(title="Построение зависимости итерации от погрешности",
                  xaxis_title="Значение погрешности под логарифмом",
                  yaxis_title="Значение итерации")

fig.show()

In [ ]:
# Вычислим оценку погрешности
r = f2(0)*((sp_x[-1] - sp_x[-2])**2) / (2*f1(0))
r

2.8335351268607296e-07